lib/SimpleHttpClient/SslClientConnection.cpp

```diff
@@ -343,15 +377,44 @@ bool SslClientConnection::connectSocket() {
 
   ERR_clear_error();
 
-  int ret = SSL_connect(_ssl);
+  int ret = -1;
+  int errorDetail = -1;
+
+  if (_isSocketNonBlocking) {
+    auto start = std::chrono::steady_clock::now();
+    while ((ret = SSL_connect(_ssl)) == -1) {
+      errorDetail = SSL_get_error(_ssl, ret);
+      if (_isInterrupted) {
+        break;
+      }
+      auto end = std::chrono::steady_clock::now();
+      if ((errorDetail != SSL_ERROR_WANT_READ &&
+           errorDetail != SSL_ERROR_WANT_WRITE) ||
+          std::chrono::duration_cast<std::chrono::seconds>(end - start)
+                  .count() >= _connectTimeout) {
+        break;
+      }
+      std::this_thread::sleep_for(std::chrono::milliseconds(50));
+    }
+  } else {
+    if ((ret = SSL_connect(_ssl)) == -1) {
+      errorDetail = SSL_get_error(_ssl, ret);
+    }
+  }
+
+  if (_isSocketNonBlocking) {
+    if (!cleanUpSocketFlags()) {
+      disconnectSocket();
+      _isConnected = false;
+      return false;
+    }
+  }
 
   if (ret != 1) {
     _errorDetails.append("SSL: during SSL_connect: ");
 
-    int errorDetail;
     long certError;
 
-    errorDetail = SSL_get_error(_ssl, ret);
     if ((errorDetail == SSL_ERROR_WANT_READ) ||
```

Comment:
If in non-blocking mode we will reach this line by the timeout - errorDetail would be exactly SSL_ERROR_WANT_READ/WRITE  and we will return true. But connection might be still not established. Is this expected?